client/gutenberg/editor/components/post-publish-button-or-toggle/index.jsx

```diff
@@ -0,0 +1,225 @@
+			unlockPostSaving,
+			createWarningNotice,
+			setState,
+			verificationNoticeLabel,
+		} = this.props;
+
+		const lockName = 'blockEditorPostSavingLock';
+
+		if ( userNeedsVerification ) {
+			lockPostSaving( lockName );
+
+			createWarningNotice( verificationNoticeLabel, {
+				id: 'verify-email-notice',
+				isDismissible: false,
+				actions: [
+					{
+						label: translate( 'Learn More' ),
+						url: '#',
+						onClick: e => {
+							e.preventDefault();
+							setState( { showEmailVerificationNotice: true } );
+						},
+					},
+				],
+			} );
+		} else {
+			unlockPostSaving( lockName );
+		}
+	}
+
+	closeVerifyEmailDialog = () => {
+		this.props.setState( { showEmailVerificationNotice: false } );
+	};
+
+	render() {
+		const {
+			forceIsDirty,
+			forceIsSaving,
+			hasPublishAction,
+			isBeingScheduled,
+			isLessThanMediumViewport,
+			isPending,
+			isPublished,
+			isPublishSidebarEnabled,
+			isPublishSidebarOpened,
+			isScheduled,
+			togglePublishSidebar,
+			showEmailVerificationNotice,
+		} = this.props;
+
+		const IS_TOGGLE = 'toggle';
+		const IS_BUTTON = 'button';
+		let component;
+
+		/**
+		 * Conditions to show a BUTTON (publish directly) or a TOGGLE (open publish sidebar):
+		 *
+		 * 1) We want to show a BUTTON when the post status is at the _final stage_
+		 * for a particular role (see https://codex.wordpress.org/Post_Status):
+		 *
+		 * - is published
+		 * - is scheduled to be published
+		 * - is pending and can't be published (but only for viewports >= medium).
+		 * 	 Originally, we considered showing a button for pending posts that couldn't be published
+		 * 	 (for example, for an author with the contributor role). Some languages can have
+		 * 	 long translations for "Submit for review", so given the lack of UI real estate available
+		 * 	 we decided to take into account the viewport in that case.
+		 *  	 See: https://github.com/WordPress/gutenberg/issues/10475
+		 *
+		 * 2) Then, in small viewports, we'll show a TOGGLE.
+		 *
+		 * 3) Finally, we'll use the publish sidebar status to decide:
+		 *
+		 * - if it is enabled, we show a TOGGLE
+		 * - if it is disabled, we show a BUTTON
+		 */
+		if (
+			isPublished ||
+			( isScheduled && isBeingScheduled ) ||
+			( isPending && ! hasPublishAction && ! isLessThanMediumViewport )
+		) {
+			component = IS_BUTTON;
+		} else if ( isLessThanMediumViewport ) {
+			component = IS_TOGGLE;
+		} else if ( isPublishSidebarEnabled ) {
+			component = IS_TOGGLE;
+		} else {
+			component = IS_BUTTON;
+		}
+
+		return (
+			<Fragment>
+				<PostPublishButton
+					forceIsDirty={ forceIsDirty }
+					forceIsSaving={ forceIsSaving }
+					isOpen={ isPublishSidebarOpened }
+					isToggle={ component === IS_TOGGLE }
+					onToggle={ togglePublishSidebar }
+				/>
+				{ showEmailVerificationNotice && (
+					<VerifyEmailDialog onClose={ this.closeVerifyEmailDialog } />
+				) }
+			</Fragment>
+		);
+	}
+}
+
+const applyWithSelect = withSelect( select => {
```

Comment:
I think this change is unnecessary compared to the original Gutenberg component.

We don't need to translate `verificationNoticeLabel` every time the state changes, but instead we should only render it translated when strictly required.
We usually do this in a component method, e.g.
```jsx
getVerificationNoticeLabel = () => {
  const { isPublished, isBeingScheduled, translate } = this.props;
  if ( isPublished ) return translate( '...' );
  else if ( isBeingScheduled ) return translate( '...' );
  return translate( '...' );
};
```

This means that in `componentDidUpdate` you should be more precise in when calling `this.togglePostSaving();`

As a side note, you might have notice that I'm suggesting to use `translate` as a prop.
Such prop is created by the `localize()` higher order component, and it's the [suggested way of using i18n-calypso](https://github.com/Automattic/i18n-calypso/blob/9556007bd2dd6d3bb88edfd0c5e8e4275edf06b4/README.md).

Reply:
To be more clear, I'd move the notice label localization into a component method, and restore the original `withSelect` altogether:

```jsx

	withSelect( select => ( {
		hasPublishAction: get(
			select( 'core/editor' ).getCurrentPost(),
			[ '_links', 'wp:action-publish' ],
			false
		),
		isBeingScheduled: select( 'core/editor' ).isEditedPostBeingScheduled(),
		isPending: select( 'core/editor' ).isCurrentPostPending(),
		isPublished: select( 'core/editor' ).isCurrentPostPublished(),
		isPublishSidebarEnabled: select( 'core/editor' ).isPublishSidebarEnabled(),
		isPublishSidebarOpened: select( 'core/edit-post' ).isPublishSidebarOpened(),
		isScheduled: select( 'core/editor' ).isCurrentPostScheduled(),
	} ) ),
```

Reply:
In fact, I had an implementation using this approach in some moment, but decided to change it exactly for what you say about `componentDidUpdate`:

> This means that in `componentDidUpdate` you should be more precise in when calling `this.togglePostSaving();``

It ended up being a bit more complex, checking 3 props (`userNeedsVerification`, `isPublished` and `isBeingScheduled`) instead of 2 (`userNeedsVerification` and `verificationNoticeLabel`).

But this is actually making more complex the `withSelect`, so it makes sense to revert it since that will also make possible to use the `localize` HOC.

Done in d5d6164.